Disable formatting of CPM.cmake file

jecassis opened the pull request:
To better support the recommended method of adding and updating CPM.cmake, it might be helpful to disable that file from formatting.

In this pull request, formatting of CPM.cmake is skipped by default. The reasoning is that I don't know if adding a new option to the interface is acceptable or simplicity is preferable.

An alternative implementation would be to filter through Git's pathspec instead. Example:
```bash
${GIT_PROGRAM} ls-files --cached --exclude-standard ${_GIT_REPOSITORY_DIR} -- ':!:**/CPM.cmake'
```

TheLartians commented:
Hey, thanks for the suggestion and PR! I definitely agree that Format.cmake should not format CPM.cmake, or any other third-party libraries for that matter. Tbh I don't think excluding a single file will be a lot of help, as users often have many other third-party scripts in their `cmake` directory that should be exempt from formatting as well.
The way this is usually dealt with in C++ is to add a `.clang_format` that disables formatting in any directory containing third-party code, which would be a good solution here as well imo. Unfortunately, [until recently](https://github.com/cheshirekow/cmake_format/issues/207) cmake_format didn't yet support such an option, but it should be included in the next release (v0.6.11). This will allow to users to simply add a `.cmake-format` configuration file in the `cmake` directory to disable formatting of any third-party CMake code.

TheLartians commented:
Alternatively, I can see that a user-defined exclude pattern might be useful and a valid workaround until cmake-format is updated. Something like this would allow users to specify single files or directories to exclude on a per-project basis:

```cmake
CPMAddPackage(
  NAME Format.cmake
  VERSION 1.5
  GITHUB_REPOSITORY TheLartians/Format.cmake
  # enable cmake formatting
  OPTIONS 
    "FORMAT_CHECK_CMAKE ON"
    "FORMAT_EXCLUDE_CMAKE (^|/)CPM\\.cmake$"
)
```

jecassis commented:
I like your idea and implemented it because I don't like how `.cmake-format` lacks flexibility. I would have liked for it to be closer to `.gitignore` or `.prettierignore`: per repo and/or subdirectory with file granularity. I tested the changes for my use cases and works as expected. Caveat: I was only able to test Ninja (Windows 10 VS2019) and Unix Makefiles (Ubuntu 20.04) as generators to work correctly as of right now when using `$` in the pattern due to escaping.

jecassis commented on CMakeLists.txt:
```diff
@@ -1,4 +1,5 @@
 option(FORMAT_CHECK_CMAKE "Enable CMake formatting.")
+option(FORMAT_CMAKE_EXCLUDE "Regular expression file exclusion pattern.")
```
Appreciate the comment. Updated.

jecassis commented:
Firstly I apologize profusely for clobbering your CI pipeline. What I thought was a quick fix ended up taking many attempts to get a workable cross-platform solution :(

On the positive side, I updated from `option` to `set` and added a CI test. The variable interpolation behaves differently on the various platforms.

**Notes**: File paths are relative to `CMAKE_SOURCE_DIR` (in this repo it is `test/`) and on Windows with MSBuild special characters are replaced such as `^`, `/` and `\\` as it thinks the `add_custom_command` arguments are all paths. So my recommendation would be to avoid them on that specific combination. Strangely using a different generator on Windows such as Ninja works identically to Unix.

jecassis commented on CMakeLists.txt:
```diff
@@ -38,10 +39,13 @@ endif()
 
 if (GIT_PROGRAM AND CMAKE_FORMAT_PROGRAM)
   function (add_cmake_format_target name)
+    if (CMAKE_FORMAT_EXCLUDE AND ((CMAKE_GENERATOR STREQUAL "Unix Makefiles") OR (CMAKE_GENERATOR STREQUAL "Ninja")))
```
Initially thought it would be better to be explicit with the tools that were tested, but I neglected to think about the rest of the ecosystems. Made the change.

TheLartians commented:
Do you think it's possible to escape special characters on Windows to allow using the same regexes regardless of the platform / generator?

jecassis commented:
Here are the takeaways:
- Importantly this works in my repositories with no special OS handling. Tested on Windows 10 and Ubuntu 20.04 inside IDEs and directly in the shells:
```
CPMAddPackage(
  NAME Format.cmake
  GITHUB_REPOSITORY jecassis/Format.cmake
  GIT_TAG f32d340004ec7bf8da1cfe81298a112a8d600e6c
  OPTIONS "FORMAT_CHECK_CMAKE ON" "CMAKE_FORMAT_EXCLUDE (^|/)CPM\\.cmake$"
)
```
- One change is really needed to make it cross-platform. It was to call `cmake-format.cmake` using the `VERBATIM` option so as to not need any special handling of the passed variable(s).
- Take note that shells and OS combinations will do pattern substitution differently. What gave the most grief was that Bash on the Windows GitHub machines was substituting the `(^|/)sample\\.cmake$` pattern with `(^|C:/Program Files/Git/)sample/.cmake$`. This was not reproducible on my local machine. So fell back on using the default which is PowerShell--which does not substitute `\\` for `\` like Bash does so only one backslash :)

Does that make it less or more confusing?

TheLartians commented:
Good find with the `VERBATIM` option! It's great that the CMake targets now behave identically on all systems.
As for the windows bash shell replacing arguments with paths: that's unfortunate, but definitely out of our control, as I assume it would behave that way with all programs.
I'd say this is good to merge. Thanks for adding this useful feature! :)

jecassis commented:
Thanks for your comments!

TheLartians commented:
The option is now released in version 1.6 🎉
